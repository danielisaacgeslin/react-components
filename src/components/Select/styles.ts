import { StyleSheet } from 'aphrodite/no-important';
import { Color } from 'utils/dist/models/general';
import { hexToRgba } from 'utils/dist/utils';

export default StyleSheet.create({
  container: {
    width: '100%',
    color: Color.BLACK
  },
  selectedContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    cursor: 'pointer',
    border: `1px solid ${hexToRgba(Color.GRAY, 50)}`,
    borderRadius: 5,
    paddingTop: 10,
    paddingBottom: 10,
    paddingLeft: 15,
    paddingRight: 15
  },
  optionListContainer: {
    display: 'flex',
    flexDirection: 'column',
    overflow: 'auto',
    margin: '0 -15px',
    backgroundColor: Color.WHITE
  },
  optionContainer: {
    paddingTop: 10,
    paddingBottom: 10,
    paddingLeft: 15,
    paddingRight: 15,
    cursor: 'pointer',
    transition: 'all 0.1s',
    ':hover': {
      backgroundColor: hexToRgba(Color.BLACK, 7)
    }
  },
  active: {
    backgroundColor: hexToRgba(Color.BLACK, 11)
  }
});
