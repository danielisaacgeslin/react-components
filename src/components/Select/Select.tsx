import React, { memo, useCallback, useState, useRef } from 'react';
import { css } from 'aphrodite/no-important';

import Icon from '../Icon';
import Dropdown from '../Dropdown';
import styles from './styles';
import { ClickOutsideListener } from '../index';

export interface ISelectProps {
  value: any;
  optionList: any[];
  maxHeight?: number;
  onChange: (value: any) => void;
  keyExtractor: (option: any) => string | number;
  renderSelected: (option: any) => JSX.Element;
  renderOption: (option: any) => JSX.Element;
}

const Select = ({ value, optionList, maxHeight = 350, onChange, keyExtractor, renderSelected, renderOption }: ISelectProps) => {
  const [isOpen, setOpen] = useState<boolean>(false);
  const open = useCallback(() => setOpen(true), [setOpen]);
  const close = useCallback(() => setOpen(false), [setOpen]);

  const ref = useRef<HTMLDivElement>();

  const onSelect = useCallback(
    option => () => {
      onChange(option);
      close();
    },
    [onChange, close]
  );

  return (
    <div ref={ref} key="Select" className={css(styles.container)}>
      <ClickOutsideListener onClickOutside={close}>
        <div data-testid="select" className={css(styles.selectedContainer)} onClick={open}>
          <div>{renderSelected(value)}</div>
          <Icon icon="angle-down" />
        </div>
        <Dropdown isOpen={isOpen}>
          <div className={css(styles.optionListContainer)} style={{ width: ref.current && ref.current.offsetWidth, maxHeight }}>
            {optionList.map(option => (
              <div
                data-testid="select-option"
                className={css([styles.optionContainer, value === option && styles.active])}
                key={keyExtractor(option)}
                onClick={onSelect(option)}
              >
                {renderOption(option)}
              </div>
            ))}
          </div>
        </Dropdown>
      </ClickOutsideListener>
    </div>
  );
};

export default memo(Select);
