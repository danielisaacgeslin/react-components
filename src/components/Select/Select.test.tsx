import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';

import Select, { ISelectProps } from './Select';

describe('Select Component', () => {
  let Component;
  let props: ISelectProps;
  let optionList;

  beforeEach(() => {
    global.console.error = () => {
      /** */
    };
    optionList = [
      { value: 1, title: 'one', description: 'this is the number number one' },
      { value: 2, title: 'two', description: 'this is the number number two' }
    ];
    props = {
      value: null,
      optionList: optionList,
      keyExtractor: option => option.value,
      renderOption: option => (
        <div>
          <div>{option.title}</div>
          <small>{option.description}</small>
        </div>
      ),
      renderSelected: option => <div>{option ? option.title : 'please select an option'}</div>,
      onChange: jest.fn()
    };
    Component = mount(<Select {...props} />);
  });

  it('should render closed', () => {
    expect(create(Component).toJSON()).toMatchSnapshot();
  });

  it('should render open', () => {
    Component.find('[data-testid="select"]').simulate('click');
    expect(create(Component).toJSON()).toMatchSnapshot();
  });

  it('should with an option selected', () => {
    props.value = optionList[0];
    Component = mount(<Select {...props} />);
    Component.find('[data-testid="select"]').simulate('click');
    expect(create(Component).toJSON()).toMatchSnapshot();
  });

  it('should select and close', () => {
    Component.find('[data-testid="select"]').simulate('click');
    Component.find('[data-testid="select-option"]')
      .first()
      .simulate('click');

    expect(props.onChange).toBeCalledWith(optionList[0]);
    expect(create(Component).toJSON()).toMatchSnapshot();
  });
});
