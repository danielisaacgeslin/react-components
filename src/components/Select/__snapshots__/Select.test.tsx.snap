// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component should render closed 1`] = `
.container {
  width: 100%;
  color: #000;
}

.selectedContainer {
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  align-items: center;
  cursor: pointer;
  border: 1px solid rgba(213, 222, 241, 0.5);
  border-radius: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 15px;
  padding-right: 15px;
}

.container {
  position: absolute;
  box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  background-color: #FFF;
  z-index: 200;
  border-radius: 3px;
  padding: 10px 15px;
  opacity: 0;
  color: #D5DEF1;
  font-size: 15px;
}

<div
  className="container"
>
  <div>
    <div
      className="selectedContainer"
      data-testid="select"
      onClick={[Function]}
    >
      <div>
        <div>
          please select an option
        </div>
      </div>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down fa-w-10 "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="container"
      onClick={[Function]}
      style={
        Object {
          "bottom": undefined,
          "left": undefined,
          "right": undefined,
          "top": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`Select Component should render open 1`] = `
.container {
  width: 100%;
  color: #000;
}

.selectedContainer {
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  align-items: center;
  cursor: pointer;
  border: 1px solid rgba(213, 222, 241, 0.5);
  border-radius: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 15px;
  padding-right: 15px;
}

.container {
  position: absolute;
  box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  background-color: #FFF;
  z-index: 200;
  border-radius: 3px;
  padding: 10px 15px;
  opacity: 0;
  color: #D5DEF1;
  font-size: 15px;
}

<div
  className="container"
>
  <div>
    <div
      className="selectedContainer"
      data-testid="select"
      onClick={[Function]}
    >
      <div>
        <div>
          please select an option
        </div>
      </div>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down fa-w-10 "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="container"
      onClick={[Function]}
      style={
        Object {
          "bottom": undefined,
          "left": undefined,
          "right": undefined,
          "top": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`Select Component should select and close 1`] = `
.container {
  width: 100%;
  color: #000;
}

.selectedContainer {
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  align-items: center;
  cursor: pointer;
  border: 1px solid rgba(213, 222, 241, 0.5);
  border-radius: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 15px;
  padding-right: 15px;
}

.container {
  position: absolute;
  box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  background-color: #FFF;
  z-index: 200;
  border-radius: 3px;
  padding: 10px 15px;
  opacity: 0;
  color: #D5DEF1;
  font-size: 15px;
}

<div
  className="container"
>
  <div>
    <div
      className="selectedContainer"
      data-testid="select"
      onClick={[Function]}
    >
      <div>
        <div>
          please select an option
        </div>
      </div>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down fa-w-10 "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="container"
      onClick={[Function]}
      style={
        Object {
          "bottom": undefined,
          "left": undefined,
          "right": undefined,
          "top": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`Select Component should with an option selected 1`] = `
.container {
  width: 100%;
  color: #000;
}

.selectedContainer {
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  align-items: center;
  cursor: pointer;
  border: 1px solid rgba(213, 222, 241, 0.5);
  border-radius: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
  padding-left: 15px;
  padding-right: 15px;
}

.container {
  position: absolute;
  box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  background-color: #FFF;
  z-index: 200;
  border-radius: 3px;
  padding: 10px 15px;
  opacity: 0;
  color: #D5DEF1;
  font-size: 15px;
}

<div
  className="container"
>
  <div>
    <div
      className="selectedContainer"
      data-testid="select"
      onClick={[Function]}
    >
      <div>
        <div>
          one
        </div>
      </div>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down fa-w-10 "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="container"
      onClick={[Function]}
      style={
        Object {
          "bottom": undefined,
          "left": undefined,
          "right": undefined,
          "top": undefined,
        }
      }
    />
  </div>
</div>
`;
